modin/experimental/xgboost/xgboost.py

```diff
@@ -0,0 +1,499 @@
+
+    Notes
+    -----
+    Currently ModinDMatrix supports only `data` and `label` parameters.
+    """
+
+    def __init__(self, data, label):
+        self.data = data
+        self.label = label
+
+    def __iter__(self):
+        yield self.data
+        yield self.label
+
+
+def create_actors(num_cpus=1, nthread=cpu_count()):
+    num_nodes = len(ray.nodes())
+
+    # Create remote actors
+    actors = [
+        ModinXGBoostNodeActor.options(num_cpus=num_cpus, resources={node: 1.0}).remote(
+            node, nthread=nthread
+        )
+        for node in ray.cluster_resources()
+        if "node" in node
+    ]
+
+    assert num_nodes == len(
+        actors
+    ), f"Number of nodes {num_nodes} is not equal to number of actors {len(actors)}."
+
+    return actors
+
+
+def _split_data_across_actors(
+    actors, set_func, X_parts, y_parts=None, evenly_data_distribution=True
+):
+    X_parts_by_actors = _set_row_partitions_by_actors(
+        actors, X_parts, evenly_data_distribution=evenly_data_distribution
+    )
+    if y_parts is not None:
+        y_parts_by_actors = _set_row_partitions_by_actors(
+            actors,
+            y_parts,
+            X_parts_by_actors,
+            evenly_data_distribution=evenly_data_distribution,
+        )
+
+    # Need to add assert on order of actors in case evenly_data_distribution=False
+    for i, actor in enumerate(actors):
+
+        X_parts = (
+            X_parts_by_actors[i]
+            if evenly_data_distribution
+            else X_parts_by_actors[i][2]
+        )
+        if y_parts is None:
+            set_func(actor, *X_parts)
+        else:
+            y_parts = (
+                y_parts_by_actors[i]
+                if evenly_data_distribution
+                else y_parts_by_actors[i][2]
+            )
+            set_func(actor, *(X_parts + y_parts))
+
+
+def _set_row_partitions_by_actors(
+    actors, row_partitions, data_for_aligning=None, evenly_data_distribution=True
+):
+
+    row_partitions_by_actors = []
+    if evenly_data_distribution:
+        num_actors = len(actors)
+        row_parts_last_idx = (
+            len(row_partitions) // num_actors
+            if len(row_partitions) % num_actors == 0
+            else len(row_partitions) // num_actors + 1
+        )
+        start_idx = 0
+        for actor in actors:
+            idx_slice = (
+                slice(start_idx, start_idx + row_parts_last_idx)
+                if start_idx + row_parts_last_idx < len(row_partitions)
+                else slice(start_idx, len(row_partitions))
+            )
+            row_partitions_by_actors.append(row_partitions[idx_slice])
+            start_idx += row_parts_last_idx
+    else:
+        actors_ips = [ray.get(actor.get_actor_ip.remote()) for actor in actors]
+        partitions_ips = [ray.get(row[0]) for row in row_partitions]
+
+        # If one of actors isn't contain partitions we make evenly
+        # distribution of data by nodes.
```

Comment:
why?.. if I have 1000 actors, and 999 have evenly spaced data, it doesn't make much sense to redistribute it all IMHO... this should probably be checking that actor IPs match with `data_for_aligning` instead of this

Reply:
It's a good point.

Reply:
It was fixed. Empty actors won't used in computations but users will notified using warning message about this.

Reply:
Fixed by next way: data will be redistributed in case when 10% or more number of nodes are empty.